### ISSUE TITLE
2.2.5 possible bug in ssl crt-list socket commands

### ISSUE THREAD

**Issue by domdom82** [haproxy/haproxy#1004]
## Detailed description of the problem

I am trying to implement a dynamic certificate updater for my crt-list in HAproxy 2.2.5.
I have noticed that somehow, when I update an existing certificate and add it to the crt-list twice, I can never remove it again.
It seems as though there is an internal reference to the certificate still kept after removal and the cert stays in use.

## Expected behavior

I would expect HAproxy to allow me to remove a certificate once it is no longer referenced in any crt-list.

## Steps to reproduce the behavior

Step 1: Add a new certificate
 
```
echo "new ssl cert mycert.pem " | nc -U haproxy.sock
echo -e "set ssl cert mycert.pem <<\n$(cat mycert.pem)\n" | nc -U haproxy.sock
echo "commit ssl cert mycert.pem " | nc -U haproxy.sock
```
 
Step 2: Reference the certificate in crt-list
 
```
echo -e "add ssl crt-list my-crt-list <<\nmycert.pem [verify: none] myhost.mydomain.com\n" | nc -U haproxy.sock
```

Step 3: openssl check

``` 
openssl s_client -connect localhost:443 -servername myhost.mydomain.com
(…)
subject=… /CN=myhost.mydomain.com
 ```

So it works as expected.
 
Step 4: Update the certificate again
 
```
echo "new ssl cert mycert.pem " | nc -U haproxy.sock. (already exists)
echo -e "set ssl cert mycert.pem <<\n$(cat mycert.pem)\n" | nc -U haproxy.sock
echo "commit ssl cert mycert.pem " | nc -U haproxy.sock
```

Step 5: Add another reference in crt-list
 
```
echo -e "add ssl crt-list my-crt-list <<\nmycert.pem [verify: none] myhost.mydomain.com\n" | nc -U haproxy.sock
```

There are now 2 references in the crt-list:
mycert.pem:7 [verify none] myhost.mydomain.com
mycert.pem:8 [verify none] myhost.mydomain.com
 
Step 6: Remove the old reference

``` 
echo "del ssl crt-list my-crt-list mycert.pem:7 " | nc -U haproxy.sock
```

This was my original idea of how to update certificates without a downtime.
 
However, I found an odd issue when trying to finally delete the crt-list entry and the certificate thereafter:
 
Step 7: Remove the last reference

``` 
echo "del ssl crt-list my-crt-list mycert.pem:8 " | nc -U haproxy.sock
```

The crt-list now contains no more references to the certificate.
 
Step 8: Delete certificate

``` 
echo "del ssl cert mycert.pem" | nc -U haproxy.sock
Can't remove the certificate: certificate 'mycert.pem' in use, can't be deleted!
```

Huh? Checking certificate status:
 
```
echo "show ssl cert mycert.pem" | nc -U haproxy.sock
Filename: mycert.pem
Status: Used
```

So I can’t delete the cert even though there is no reference to it.
But there seems to be still a “hidden” reference somehow, because I can still see it in TLS:
 
Step 9: openssl check again

```
openssl s_client -connect localhost:443 -servername myhost.mydomain.com
(…)
subject=… /CN=myhost.mydomain.com
```

This should not be. The crt-list no longer contains any SNI filter for this domain, so it should fall back to the default certificate.

## Do you have any idea what may have caused this?

When adding the crt-list ref a second time it seems there is something going wrong with the internal ref count.
I assume this because when I only add the cert once and then remove it again from the crt-list, I can remove the certificate successfully.

## Do you have an idea how to solve the issue?

Check what's happening in the logic between `(new|set|del) ssl cert` and `(add|del) ssl crt-list`

## What is your configuration?

```
global
    log /run/log-haproxy len 4096 format rfc3164 syslog info
    daemon
    nbthread 4
    user vcap
    group vcap
    maxconn 512000
    spread-checks 4
    lua-load /var/vcap/packages/haproxy-config/reverse-proxy.lua
    lua-load /var/vcap/packages/haproxy-config/gbaas-deny.lua
    lua-load /var/vcap/packages/haproxy-config/health-check.lua
    tune.ssl.default-dh-param 2048
    tune.bufsize 35840
    tune.maxrewrite 4096
    stats socket /var/vcap/sys/run/haproxy/stats.sock mode 600 expose-fd listeners level admin
    stats timeout 2m
    ssl-default-bind-options no-sslv3 no-tlsv10 no-tlsv11
    ssl-default-bind-ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-SHA256:ECDHE-RSA-AES256-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES128-SHA256:DHE-RSA-AES256-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-RSA-AES256-SHA:AES128-SHA256:AES128-SHA
    ssl-default-server-options no-sslv3 no-tlsv10 no-tlsv11
    ssl-default-server-ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-SHA256:ECDHE-RSA-AES256-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES128-SHA256:DHE-RSA-AES256-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-RSA-AES256-SHA:AES128-SHA256:AES128-SHA

defaults
    log global
    option log-health-checks
    option log-separate-errors
    maxconn 512000
    option http-server-close
    option httplog
    option forwardfor
    option contstats
    timeout connect         30000ms
    timeout client          960000ms
    timeout server          960000ms
    timeout tunnel          3600000ms
    timeout http-keep-alive 60000ms
    timeout http-request    90000ms
    timeout queue           30000ms


listen stats_1
    bind localhost:9000
    acl private src 127.0.0.1/32
    http-request deny unless private
    mode http
    stats enable
    stats hide-version
    stats realm "Haproxy Statistics"
    stats uri /haproxy_stats
    stats auth haproxy_stats_user:haproxy_stats_password


listen health_check_http_url
    bind :8080
    mode http
    monitor-uri /health
    acl http-routers_down nbsrv(http-routers) eq 0
    monitor fail if http-routers_down


# HTTP Frontend {{{
frontend http-in
    mode http
      grace 15000
    bind :80
    option clitcpka
log-format "%ci:%cp [%tr] (%ID) %ft %b/%s %Th/%Ti/%TR/%Tw/%Tc/%Tr/%Tt %ST %B %CC %CS %tsc %ac/%fc/%bc/%sc/%rc %sq/%bq %hr %hs %{+Q}r %sslc %sslv %[ssl_fc_has_sni] %[ssl_c_used] %[ssl_fc_has_crt] %[ssl_c_err] %{+Q}[ssl_c_s_dn] %{+Q}[ssl_c_i_dn]"
capture request header User-agent len 256
no option http-server-close
acl haproxy-health-check hdr(host) -i haproxy-health-check.cf.cfi01.ali.cfi.sapcloud.io
http-request use-service lua.health-check if haproxy-health-check
acl blocklist src -f /var/vcap/jobs/haproxy/config/blacklist_cidrs.txt
acl gbaas-blocklist-test hdr_beg(host) -i test-gbaas-access-denied
acl allowhost hdr(host) -i -f /var/vcap/jobs/haproxy/config/hosts/allow-list
http-request use-service lua.gbaas-deny if blocklist !allowhost
http-request use-service lua.gbaas-deny if gbaas-blocklist-test
http-request lua.rp
http-request del-header X-SCP-Request-Id
unique-id-format %H-%{+X}Ts-%{+X}rt
unique-id-header X-SCP-Request-Id

    acl whitelist src -f /var/vcap/jobs/haproxy/config/whitelist_cidrs.txt
    tcp-request content accept if whitelist
    acl blacklist src -f /var/vcap/jobs/haproxy/config/blacklist_cidrs.txt
    tcp-request content reject if blacklist
    tcp-request content reject
    capture request header Host len 256
    default_backend http-routers
    acl block_test_request_deny hdr_beg(host) -i test-request-deny

    http-request deny if block_test_request_deny
    acl xfp_exists hdr_cnt(X-Forwarded-Proto) gt 0
    http-request add-header X-Forwarded-Proto "http" if ! xfp_exists
    redirect scheme https code 301 if !{ ssl_fc }
# }}}

# HTTPS Frontend {{{
frontend https-in
    mode http
      grace 15000
    bind :443  ssl crt-list /var/vcap/jobs/haproxy/config/ssl/crt-list  ca-file /etc/ssl/certs/ca-certificates.crt verify optional
    option clitcpka
log-format "%ci:%cp [%tr] (%ID) %ft %b/%s %Th/%Ti/%TR/%Tw/%Tc/%Tr/%Tt %ST %B %CC %CS %tsc %ac/%fc/%bc/%sc/%rc %sq/%bq %hr %hs %{+Q}r %sslc %sslv %[ssl_fc_has_sni] %[ssl_c_used] %[ssl_fc_has_crt] %[ssl_c_err] %{+Q}[ssl_c_s_dn] %{+Q}[ssl_c_i_dn]"
capture request header User-agent len 256
no option http-server-close
acl haproxy-health-check hdr(host) -i haproxy-health-check.cf.cfi01.ali.cfi.sapcloud.io
http-request use-service lua.health-check if haproxy-health-check
acl blocklist src -f /var/vcap/jobs/haproxy/config/blacklist_cidrs.txt
acl gbaas-blocklist-test hdr_beg(host) -i test-gbaas-access-denied
acl allowhost hdr(host) -i -f /var/vcap/jobs/haproxy/config/hosts/allow-list
http-request use-service lua.gbaas-deny if blocklist !allowhost
http-request use-service lua.gbaas-deny if gbaas-blocklist-test
http-request lua.rp
http-request del-header X-SCP-Request-Id
unique-id-format %H-%{+X}Ts-%{+X}rt
unique-id-header X-SCP-Request-Id

    acl whitelist src -f /var/vcap/jobs/haproxy/config/whitelist_cidrs.txt
    tcp-request content accept if whitelist
    acl blacklist src -f /var/vcap/jobs/haproxy/config/blacklist_cidrs.txt
    tcp-request content reject if blacklist
    tcp-request content reject
    http-request del-header X-Forwarded-Client-Cert
    http-request set-header X-Forwarded-Client-Cert %[ssl_c_der,base64] if { ssl_c_used }
    http-response set-header Strict-Transport-Security max-age=31536000;\ includeSubDomains;\ preload;
    capture request header Host len 256
    default_backend http-routers
    acl block_test_request_deny hdr_beg(host) -i test-request-deny

    http-request deny if block_test_request_deny
    acl xfp_exists hdr_cnt(X-Forwarded-Proto) gt 0
    http-request add-header X-Forwarded-Proto "https" if ! xfp_exists
# }}}


# Default Backend {{{
backend http-routers
    mode http
    balance roundrobin
    option srvtcpka
no option http-server-close
http-reuse safe
option prefer-last-server

        errorfile 503 /var/vcap/jobs/haproxy/errorfiles/custom503.http
        option httpchk GET /health

    server node0 1.2.3.4:443 check inter 1000 port 8080 ssl verify required ca-file /var/vcap/jobs/haproxy/config/backend-ca-certs.pem verifyhost my-backend.com
# }}}

# Routed Backends {{{
# }}}

# TCP Routing  {{{

# TCP Backends
frontend tcp-frontend_scheduler
    mode tcp
    bind :2222
    default_backend tcp-scheduler
      grace 15000

backend tcp-scheduler
    mode tcp

    server node0 10.1.1.68:2222 check inter 1000

listen health_check_http_tcp-scheduler
    bind :2223
    mode http
    monitor-uri /health
    acl tcp-scheduler-routers_down nbsrv(tcp-scheduler) eq 0
    monitor fail if tcp-scheduler-routers_down

# }}}
```

## Output of `haproxy -vv` and `uname -a`

```
Linux b30d1f6e-4e56-45b1-99d1-1171341f091d 4.15.0-122-generic #124~16.04.1-Ubuntu SMP Thu Oct 15 16:08:36 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

HA-Proxy version 2.2.5-34b2b10 2020/11/05 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2025.
Known bugs: http://www.haproxy.org/bugs/bugs-2.2.5.html
Running on: Linux 4.15.0-122-generic #124~16.04.1-Ubuntu SMP Thu Oct 15 16:08:36 UTC 2020 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=yes USE_STATIC_PCRE2=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE +STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=4).
Built with OpenSSL version : OpenSSL 1.0.2g  1 Mar 2016
Running on OpenSSL version : OpenSSL 1.0.2g  1 Mar 2016
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2
Built with Lua version : Lua 5.4.1
Built with network namespace support.
Built with zlib version : 1.2.8
Running on zlib version : 1.2.8
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.34 2019-11-21
PCRE2 library supports JIT : no (libpcre2 build without JIT?)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 5.4.0 20160609

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
              h2 : mode=HTTP       side=FE|BE     mux=H2
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services : none

Available filters :
	[SPOE] spoe
	[COMP] compression
	[TRACE] trace
	[CACHE] cache
	[FCGI] fcgi-app
```

## If HAProxy crashed: Last outputs and backtraces

```
haproxy did not crash
```

## Additional information (if helpful)

Initial crt-list: (4 entries + 2 empty lines = first additional cert will be inserted at line `:7`)

```
/var/vcap/jobs/haproxy/config/ssl/cert-0.pem [verify none] !*.crt-test.mydomain.com !*.cert.mydomain.com
/var/vcap/jobs/haproxy/config/ssl/cert-1.pem [ca-file /var/vcap/jobs/haproxy/config/ssl/ca-file-1.pem verify optional] *.crt-test.mydomain.com !*.cert.mydomain.com
/var/vcap/jobs/haproxy/config/ssl/cert-2.pem [ca-file /var/vcap/jobs/haproxy/config/ssl/ca-file-2.pem verify optional] !*.crt-test.mydomain.com *.cert.mydomain.com
/var/vcap/jobs/haproxy/config/ssl/cert-3.pem [verify none]
```

**Comment by wlallemand** [comment]
Looks like a bug indeed, I'm looking into this.

**Comment by wlallemand** [comment]
I can reproduce simply this way:
- having the same certificate several time in a crt-list
- updating it with "set ssl cert/commit ssl cert"
- removing it from the crt-list
- trying to delete it completely

I'm reviewing the code but I suppose the problem is in the "commit ssl cert" command.

**Comment by wlallemand** [comment]
I found the problem, the link between the instances and the crtlist entry was not created during a "commit ssl cert".
The fix above should fix the issue.

**Comment by domdom82** [comment]
hi @wlallemand thanks for the quick fix! do you know if this will make it into 2.2.7?

**Comment by wlallemand** [comment]
Sure, the fix will be available in the next 2.3 and 2.2 releases.
